### ISSUE TITLE
expire emails arent sent out

### ISSUE THREAD

**Issue by teadur** [internetee/registry#186]
When domain goes to expire status, reminder email should be sent out according to the configured 
Setting.expiration_reminder_mail parameter (defaults to 2)
So by default configuration expire reminder email should be sent out once to domain owner and all administrative contacts 2 days after domain has expired.
#188

Check expire_time (domain.valid_to = domain.valid_from.utc.beginning_of_day + convert_period_to_time(domain.period, domain.period_unit) + 1.day)

**Comment by artur-beljajev** [comment]
At what time should it happen?

**Comment by vohmar** [comment]
please check how the process worked before.

But you might solve this like this: 

same job that sets the domains in expired status creates Que job for that domain with the wait delay as set in system parameters configurable in admin interface - 'Expiration Reminder Mail' (/admin/settings).
The job must have a verification to check if the domain is still expired before sending out any emails.

**Comment by artur-beljajev** [comment]
"expired" = statuses.include?(DomainStatus::EXPIRED) ?

**Comment by kritik** [comment]
Yes, here expired domains are checked by states. This was in plans to move domain states to dynamic ones. Right now we have dynamic states only on contacts.

**Comment by artur-beljajev** [comment]
@vohmar Is this a valid subject "The xxx.xx domain has expired"?

**Comment by vohmar** [comment]
@artur-beljajev yes if xxx.xx is replaced with the actual domain name ie internet.ee

**Comment by artur-beljajev** [comment]
Needs testing on staging + mailcatcher

**Comment by vohmar** [comment]
@artur-beljajev did you test this?
```
2016-10-11 12:46:11 UTC DomainCron.start_expire_period: #86900 (valid-too.ee) {"outzone_at"=>[nil, Wed, 26 Oct 2016 00:00:00 EEST +03:00], "delete_at"=>[nil, Fri, 25 Nov 2016 02:00:00 EET +02:00], "statuses"=>[["ok"], ["ok", "expired"]]}
/home/registry/registry/releases/368/app/models/domain_cron.rb:48:in `block in start_expire_period': undefined method `days' for "2":String (NoMethodError)
        from /home/registry/registry/releases/368/vendor/bundle/ruby/2.2.0/gems/activerecord-4.2.5.2/lib/active_record/relation/delegation.rb:46:in `each'
        from /home/registry/registry/releases/368/vendor/bundle/ruby/2.2.0/gems/activerecord-4.2.5.2/lib/active_record/relation/delegation.rb:46:in `each'
        from /home/registry/registry/releases/368/app/models/domain_cron.rb:42:in `start_expire_period'
        from /home/registry/registry/releases/368/vendor/bundle/ruby/2.2.0/gems/railties-4.2.5.2/lib/rails/commands/runner.rb:62:in `<top (required)>'
        from /home/registry/registry/releases/368/vendor/bundle/ruby/2.2.0/gems/railties-4.2.5.2/lib/rails/commands/runner.rb:62:in `eval'
        from /home/registry/registry/releases/368/vendor/bundle/ruby/2.2.0/gems/railties-4.2.5.2/lib/rails/commands/runner.rb:62:in `<top (required)>'
        from /home/registry/registry/releases/368/vendor/bundle/ruby/2.2.0/gems/railties-4.2.5.2/lib/rails/commands/commands_tasks.rb:123:in `require'
        from /home/registry/registry/releases/368/vendor/bundle/ruby/2.2.0/gems/railties-4.2.5.2/lib/rails/commands/commands_tasks.rb:123:in `require_command!'
        from /home/registry/registry/releases/368/vendor/bundle/ruby/2.2.0/gems/railties-4.2.5.2/lib/rails/commands/commands_tasks.rb:90:in `runner'
        from /home/registry/registry/releases/368/vendor/bundle/ruby/2.2.0/gems/railties-4.2.5.2/lib/rails/commands/commands_tasks.rb:39:in `run_command!'
        from /home/registry/registry/releases/368/vendor/bundle/ruby/2.2.0/gems/railties-4.2.5.2/lib/rails/commands.rb:17:in `<top (required)>'
        from bin/rails:9:in `require'
        from bin/rails:9:in `<main>'
```

**Comment by artur-beljajev** [comment]
Fixed.

I covered code with tests as much as I could and taking into account time limitation. In this case we need to add specs for DomainMailer#start_expire_period method, but I would first extract that part out of this class at all.

**Comment by vohmar** [comment]
@artur-beljajev que job is created on expire, but the job does not work/run - an email is never sent and job is not removed  ie "run at: 4 minutes ago"

**Comment by artur-beljajev** [comment]
To detect the problem I need to see queue error log. Is it available on staging (you tested it on staging, right?)?

If it is not, this ticket is bloked by #211 and #207.

**Comment by vohmar** [comment]
@artur-beljajev tested this on staging. The jobs finish with no errors - failure count is 0 error log in GUI is empty.

**Comment by vohmar** [comment]
@artur-beljajev the job is flawed it just does not run. Please consult que documentation (https://github.com/chanks/que/blob/master/docs/writing_reliable_jobs.md) or take a look at other already existing jobs that do work (https://github.com/internetee/registry/tree/staging/app/jobs).

You also should add destroy function to remove the job from Que once it is finished and you can add some error logging.

**Comment by artur-beljajev** [comment]
1. "You also should add destroy function to remove the job from Que once it is finished" Are you sure?
2. "and you can add some error logging" Why? Where?

**Comment by artur-beljajev** [comment]
Needs testing on staging.

**Comment by vohmar** [comment]
deploy fails: https://travis-ci.org/internetee/registry/builds/171955036
cannot test

**Comment by artur-beljajev** [comment]
https://docs.travis-ci.com/user/notifications

**Comment by artur-beljajev** [comment]
Please do not remove branch registry-186, since I am implementing #180 there as well.

**Comment by artur-beljajev** [comment]
Blocked by #233

**Comment by artur-beljajev** [comment]
Fixes are in associated branch.

**Comment by vohmar** [comment]
@artur-beljajev tests failed
* destroy function missing - finished jobs are never removed from Que scheduled job list due to that
* logging needed to track Que actions - Que does not keep history, we use syslog to track what and when Que run

**Comment by artur-beljajev** [comment]
1. "destroy" is added to job even though the guide says it is not mandatory (https://github.com/chanks/que/blob/master/docs/writing_reliable_jobs.md).
2. I don't think any logging is necessary here

**Comment by vohmar** [comment]
@artur-beljajev did you test this before you released it to staging? Still no expire emails, no que jobs are generated.
Logging is still necessary - for example in case of temporary Que failure a job fails, is re-run automatically and passes later, job is deleted with no sign left behind. How do you track that a job failed number of times before successful run? We log everything that changes anything about domain or related data and is associated with this - automatic notices to domain contacts and domain deletions are among those actions.

**Comment by artur-beljajev** [comment]
1. You probably forgot to merge it with staging
2. Let's specify explicitly, what needs to be logged

**Comment by vohmar** [comment]
que job error:
```
wrong number of arguments (1 for 0)
/home/registry/registry/releases/396/app/jobs/domain_expiration_email_job.rb:2:in `run'
/home/registry/registry/releases/396/vendor/bundle/ruby/2.2.0/gems/que-0.10.0/lib/que/job.rb:15:in `_run'
/home/registry/registry/releases/396/vendor/bundle/ruby/2.2.0/gems/que-0.10.0/lib/que/job.rb:99:in `block in work'
/home/registry/registry/releases/396/vendor/bundle/ruby/2.2.0/gems/que-0.10.0/lib/que/adapters/active_record.rb:5:in `block in checkout'
/home/registry/registry/releases/396/vendor/bundle/ruby/2.2.0/gems/activerecord-4.2.5.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:292:in `with_connection'
/home/registry/registry/releases/396/vendor/bundle/ruby/2.2.0/gems/que-0.10.0/lib/que/adapters/active_record.rb:34:in `checkout_activerecord_adapter'
/home/registry/registry/releases/396/vendor/bundle/ruby/2.2.0/gems/que-0.10.0/lib/que/adapters/active_record.rb:5:in `checkout'
/home/registry/registry/releases/396/vendor/bundle/ruby/2.2.0/gems/que-0.10.0/lib/que/job.rb:82:in `work'
/home/registry/registry/releases/396/vendor/bundle/ruby/2.2.0/gems/que-0.10.0/lib/que/worker.rb:78:in `block in work_loop'
/home/registry/registry/releases/396/vendor/bundle/ruby/2.2.0/gems/que-0.10.0/lib/que/worker.rb:73:in `loop'
/home/registry/registry/releases/396/vendor/bundle/ruby/2.2.0/gems/que-0.10.0/lib/que/worker.rb:73:in `work_loop'
/home/registry/registry/releases/396/vendor/bundle/ruby/2.2.0/gems/que-0.10.0/lib/que/worker.rb:17:in `block in initialize'
```

que should log function run, domain id, result and timestamp at least. In the case described above no error was saved to log.

**Comment by artur-beljajev** [comment]
Fixed

**Comment by artur-beljajev** [comment]
Emails with the following subjects need to be tested:
- "Domeeni arturi-domeen.ee registreerija vahetuse taotlus on tühistatud / arturi-domeen.ee registrant change cancelled"
- "Domeeni arturi-domeen.ee registreerija vahetuse taotlus tagasi lükatud / arturi-domeen.ee registrant change declined"
- "Domeeni arturi-domeen.ee registreerija vahetus protseduur on algatatud / arturi-domeen.ee registrant change"
- "Kinnitustaotlus domeeni arturi-domeen.ee registreerija vahetuseks / Application for approval for registrant change of arturi-domeen.ee"
- "Kustutusmenetluse teade"
- "The kaalukoda.ee domain has expired"
- "Kinnitustaotlus domeeni arturi-domeen.ee kustutamiseks .ee registrist / Application for approval for deletion of arturi-domeen.ee"

**Comment by vohmar** [comment]
Expire email tests:
- Once again Que must log time, Job, Args, success/error, error message in case of error
- please add check to make sure two addresses the expire mail is sent to are not the same - currently if registrant and admin contact are the same the contact receives two expire messages (this goes for any email sent out - each message should be sent only once to any email address)
- email addresses should be validated before email is sent, invalid addresses removed and logged. For example this is the address registry tried to send exp e-mails for a domain with invalid adm e-mail address: <admin(at)test.ee, registrant@test.ee>. So in this case noone will receive the message.

**Comment by vohmar** [comment]
Other failed tests:
* no e-mail is sent on expired pending registrant change ("Domeeni arturi-domeen.ee registreerija vahetuse taotlus on tühistatud / arturi-domeen.ee registrant change cancelled")

minor issues with message formats and wording (ie: fd email subject only in Estonian; notice.text "ei anna kinnitust enne <%= Setting.expire_pending_confirmation %> tundi") - will solve these in separate ticket.

**Comment by artur-beljajev** [comment]
1. As I mentioned before, I don't think any logging is necessary here. There is a config key named "config.action_mailer.raise_delivery_errors" which is responsible for throwing an exception if sending fails and it is set to true for all environments. #208 can solve error logging in prod by installing errbit
2. Tried to reproduce, but was unable to. http://prntscr.com/d8wmfn I got only one single email
3. Removed from where? Probably, an exception will be thrown in case of invalid email (see no. 1). In general, email format validation should happen before saving it to the database and currently there is a check for this. If by validation you mean email existence, then it is different topic which should be handled out of this ticket's context. Maybe activemailer already does this, I haven't checked.

**Comment by artur-beljajev** [comment]
4. I am unable to reproduce it, I got the email (RegistrantChangeMailer#expired mailer).

**Comment by vohmar** [comment]
1. requirements for logging sent by sysadmins
2. nr of emails sent is indicated by the number of recipient addresses in the mail heading in Mailcather
3. this is validation thing - ideally there are no invalid addresses in the system, but in reality there are. So lets not try to send emails to invalid strings. As you might have noticed in the example  "<admin(at)test.ee, registrant@test.ee>" noone will receive any emails.
4. to reproduce please send a domain:delete EPP query to registry, click on the link sent to the registrant's email address and confirm the request by clicking on the "confirm domain delete" button. As a result domain is set to pendingDelete state but no email notification.

**Comment by artur-beljajev** [comment]
1. Logging is added
2. Fixed
3. Invalid emails lead to exception, so I think there is no need for additional check
4. Still unable to reproduce. This is what I got http://prntscr.com/dalt5p. Please specify the expected behaviour.

**Comment by artur-beljajev** [comment]
3. Invalid emails are now handled as well

**Comment by vohmar** [comment]
in production, closed, branch deleted
